### ISSUE TITLE
Seems like you are running VS Code in a remote environment (VS Code running remote docker container)

### ISSUE THREAD

**Issue by malanciault** [pieces-app/support#550]
### Software

VS Code

### Operating System / Platform

macOS

### Your Pieces OS Version

11.02

### Early Access Program

- [ ] Yes, this is related to an Early Access Program feature.

### Kindly describe the bug and include as much detail as possible on what you were doing so we can reproduce the bug.

I'm having the same problem as this: https://github.com/pieces-app/support/issues/269

Using VS Code project within a docker container and getting this message:

> Woops! It seems like you are running VS Code in a remote environment & Pieces for VS Code cannot connect to Pieces OS. Please refer to our documentation on how to resolve this issue.

![Image](https://github.com/user-attachments/assets/ffeb53fd-0787-4bc2-9faf-c5f3d179294b)

Pieces OS version:

![Image](https://github.com/user-attachments/assets/640d35af-6106-4717-9269-c51205ff23fb)

I followed this guide but with no luck: https://docs.pieces.app/product-highlights-and-benefits/connecting-to-a-remote-pieces-os-instance

LAN IP:

<img width="567" alt="Image" src="https://github.com/user-attachments/assets/1dfb849d-0764-4ae0-801c-c55f9b3e5d7e" />

VS Code settings, pieces custom URL:

![Image](https://github.com/user-attachments/assets/aaed7eb8-36e7-49cb-b6e0-84271adb8027)

What else can I try?

Thanks!

**Comment by malanciault** [comment]
To add more info, even when I have no VS Code project opened, I still get the same error message, so it is not about remote docker...

![Image](https://github.com/user-attachments/assets/7a644955-7433-4439-907a-4b50b259b98b)

**Comment by robert-at-pieces** [comment]
Hello @malanciault. It looks like our documentation is a little out of date. Starting with Pieces OS version 11.0.0 the default port is no longer 1000. Pieces OS now tries to open on port 39300. If for some reason it fails to open on that port, it will try each sequentially up to 39333.

In you case you should be able to try http://192.168.0.129:39300.

Give it a shot and keep us posted! Thanks.

**Comment by malanciault** [comment]
Still no luck

![Image](https://github.com/user-attachments/assets/3432e1ef-886a-438f-baba-e70135668f0d)

![Image](https://github.com/user-attachments/assets/3ebe3071-866d-4822-802d-33cb261bbfaf)

Anything else I could try ?

**Comment by robert-at-pieces** [comment]
How is your VSCode remote instance being hosted? Inside docker?

**Comment by malanciault** [comment]
In fact, look, I don't even think it's related to a remote instance:
![Image](https://github.com/user-attachments/assets/08c1efe2-a796-46e9-9c11-15c703b3479d)

This is what I get when I open VS Code, even before opening any project. 

Thanks for your help!

**Comment by malanciault** [comment]
Conducting more test. 

So, If I remove the Custom URL, then it seems to work before I open a project:
![Image](https://github.com/user-attachments/assets/f21e9c3e-ba7a-48eb-ba90-1f0470c875d2)

If I open a project that is not remote, it is also working :)

But If I open a remote project, then it does not work:
![Image](https://github.com/user-attachments/assets/f0dfa5ca-ae14-4357-aba4-fa9c65e9be7a)

So tho answer your question:
> How is your VSCode remote instance being hosted? Inside docker?

Yes, inside Docker.

**Comment by robert-at-pieces** [comment]
TLDR: There is an issue on our end. We will need to issue a patch to allow listening on all interfaces.

----
@malanciault Thank you for your patience. After further investigation it seems some networking changes in the latest major release are partly to blame for this issue.

The main issue identified is that, for tighter security by default, the default listening interface of Pieces Server is loopback starting from version 11.0.0.

Because you are connecting remotely from the DevContainer, the server will not listen to the address you have provided.

Docker mounts access to localhost through 'host.docker.internal' resolution from within the container.

We can see this works from the DevContainer by running the command:

```
 curl host.docker.internal:39300/.well-known/health
```
And it responds with something like:

```
ok:b454f7e3-bf2d-4ac9-b5b3-f4d8eeed5e6a
```

Unfortunately, the host.docker.internal resolution does not play well with our Webview tool windows which cannot resolve host.docker.internal.

In local testing, switching our back to listening on all interfaces resolved the problem by being able to use my LAN address in the custom url.

We will work on having a setting able to support changing this.

**Comment by malanciault** [comment]
OK thanks.

Is there something I can do on my end for it to work before a new version ?

Thanks

**Comment by robert-at-pieces** [comment]
We have a patch to the VSCode plugin currently being tested in staging. I would recommend holding off for this patch as it would be simpler. When the patch is out you should be able to remove the customUrl completely and it will detect Pieces OS automatically.

---
The workaround for this issue would be a little complex. You would have to create a dns name entry like 'pieceshost' and add it to the /etc/hosts of the container and have it resolve to 'host.docker.internal'.

Outside of the container you would have to setup this dns name to resolve to 'localhost'.

You can then specify a custom url like "http://pieceshost:39300" and it should then be able to resolve Pieces OS in the container and in the Copilot webview.

**Comment by malanciault** [comment]
Thanks Robert, I will wait for the patch. How can I be informed when the patch is released ?

Thanks!

**Comment by robert-at-pieces** [comment]
VSCode should show a new version in the extension browser when it has been released.

I will also be sure to comment on this ticket.

Looking to release the patch today or tomorrow.

**Comment by robert-at-pieces** [comment]
@malanciault version 2.0.1 of the VSCode extension is out now: https://marketplace.visualstudio.com/items?itemName=MeshIntelligentTechnologiesInc.pieces-vscode

I have tested this version on MacOS and Windows using multiple DevContainers and the connection was successful when the custom URL is blank.

Let us know if your issue has been resolved! Thanks!

**Comment by malanciault** [comment]
So, the plugin has been updated and is enabled within the container:
![Image](https://github.com/user-attachments/assets/9a15ff91-17fd-4122-b45a-2bea287c3b41)

I closed VS Code.

I ensured that Pieces OS is running
![Image](https://github.com/user-attachments/assets/63205772-2640-4935-ae4c-90c7428f2a4e)

But still:
![Image](https://github.com/user-attachments/assets/509c1dd3-4742-4074-a387-4120bc6c2987)

Let me know what I am missing! Thanks

**Comment by robert-at-pieces** [comment]
@malanciault Would you be willing to set some time up for a video call to get to the bottom of this?

**Comment by malanciault** [comment]
Of course, thanks for offering, send me an email at malanciault@illuxi.com.

Thanks again!
